Add more ways to translate nucleotide sequence with dashes

mmokrejs opened the pull request:
Currently one cannot translate sequences aligned (padded) with dashes into a protein sequence as the in-built tests are way too strict. Add a couple of arguments to the translate() functions to overcome the road blocks.

Typically padding symbols for gaps are '-' but this default can be changed on the fly if necessary.

With this change one can with respect_alignment=True translate a padded alignment file in FASTA format so that '---' translates to '-', 'AT-' translates to 'X'.

In an alternative mode with ignore_gaps=True it is possible to ignore the padding dashes so that 'AT-GATG' translates to 'MM'.

<!--- Please read each of the following items and confirm by replacing
 !--the [ ] with a [X] --->

- [X] I hereby agree to dual licence this and any previous contributions under both
the _Biopython License Agreement_ **AND** the _BSD 3-Clause License_.

- [X] I have read the ``CONTRIBUTING.rst`` file, have run ``pre-commit``
locally, and understand that continuous integration checks will be used to
confirm the Biopython unit tests and style checks pass with these changes.

- [X] I have added my name to the alphabetical contributors listings in the files
``NEWS.rst`` and ``CONTRIB.rst`` as part of this pull request, am listed
already, or do not wish to be listed. (*This acknowledgement is optional.*)

<!--- Briefly describe the changes included in this pull request below
 !--- starting with 'Closes #...' if appropriate --->

Closes #...

mdehoon commented:
How would you use this in practice?

mmokrejs commented:
Imagine you have as an input multiFASTA file, either padded or unpadded. In case of the unpadded one, due to sequencing errors often uyou have frame-breaking INSertions or DELetions.

If you align the entries you have a well aligned file but the individual entries cannot be translated because sooner or later one of those InDels breaks the reading frame. Solution is to pad them so the few INSertions get translated as 'X' but the rest is in-frame. The few DELetions get translated as 'X' too, and the rest is translated in a proper frame.

For a lazy approach, you can try to translate the entries as they are and if the translating fucntion comes across a dash it ignores it, like when you feed padded FASTA into NCBI blastn, it just skips the dashes to make life easer to the user.

The changes allowed me to overcome design flaw in gofasta, which dropped INSertions from the multiple sequence alignment: https://github.com/virus-evolution/gofasta/issues/55 . Now I do not need it at all.

The current behavior is unchanged. Probably I should have used '-' as the default for `gap` argument instead of `None`. But I tried to keep current behavior.

Example:
https://host-patho-evo.github.io/mutation_scatter_plot/
https://github.com/host-patho-evo/mutation_scatter_plot/
https://www.biorxiv.org/content/10.1101/2025.04.23.650148v1

mdehoon commented:
@mmokrejs  Is this effectively the same as removing the gap characters before translating?

If so, you could do that instead, or you could use `Bio.Align` (instead of `Bio.AlignIO`) to read the file. `Bio.Align` stores the sequences without the dashes.

peterjc commented:
> With this change one can with respect_alignment=True translate a padded alignment file in FASTA format so that '---' translates to '-', 'AT-' translates to 'X'.

Note "---" already does translate as "-":

```pycon
>>> from Bio.Seq import Seq
>>> Seq("---").translate()
Seq('-')
```

But I fail to see how translating 'AT-' to 'X' makes sense (bar the corner-case where this is the trailing right end of a partial/truncated sequence where I would suggest `.rtrim("-")` and right padding with "N" to a multiple of three).

mmokrejs commented:
> > With this change one can with respect_alignment=True translate a padded alignment file in FASTA format so that '---' translates to '-', 'AT-' translates to 'X'.
> 
> Note "---" already does translate as "-":
> 
> ```
> >>> from Bio.Seq import Seq
> >>> Seq("---").translate()
> Seq('-')
> ```
> 
> But I fail to see how translating 'AT-' to 'X' makes sense (bar the corner-case where this is the trailing right end of a partial/truncated sequence where I would suggest `.rtrim("-")` and right padding with "N" to a multiple of three).

When you parse a FASTA file with aligned sequences the dashes might appear "in the middle" of the sequence. I will improved the https://raw.githubusercontent.com/host-patho-evo/mutation_scatter_plot/refs/heads/main/tests/inputs/test.fa example with more peculiar cases. The point is that biopython raised exceptions here and there, User has now the option to override the default and let it continue.

Imagine tyou want to translate NGS reads with some errors of a protein coding region.


I will be back on Monday to answer more.

peterjc commented:
Perhaps your suggested naming `respect_alignment=True` is confusing me as it seems rather you discard the alignment information (which was saying you have a partial codon which cannot be translated).

I echo @mdehoon's question about why not use `.replace("-", "").translate()` here.

mmokrejs commented:
> Perhaps your suggested naming `respect_alignment=True` is confusing me as it seems rather you discard the alignment information (which was saying you have a partial codon which cannot be translated).
> 
> I echo @mdehoon's question about why not use `.replace("-", "").translate()` here.

That is definitelety what I am not after. I want biopython to be robust. By default, it does what it does. With `respect_alignment=True` it respects the padding. Alternatively with `ignore_gaps=True` it does `.replace("-", "").translate()` on the fly. Maybe `ignore_gaps` should be renamed to `discard_gaps` or `discard_padding`?


`respect_alignment=True` translates the sequence as it is, if there are dashes, it produces 'X' but continues. Currently, it dies.

Weird that for you '---' translated to '-' but I started with an older biopython version, 1.83 and I think 1.81 too on another host.

See https://github.com/biopython/biopython/pull/4992/commits/d6fdc48beeb6121ace9a7e4c6ea85ca3fd0d5480#diff-bdfee92a19a985d8313cb909cc673c4f02acb6d54f4aeea337b0e03e7f6b5803R2803-R2810 , the patch isn't so long.

peterjc commented:
I feel the translate method & function are already rather complex, so am reluctant to add any more optional arguments.

I think explicitly writting `.replace("-", "").translate()` is better than adding an option to for this. It seem unambigous, while adding it to the translate method almost forces the reader to consult the documentation.

For the other idea, I _might_ have suggested adding an argument about gapped partial codons defaulting to `None` meaning the current behaviour of treating partial codons as errors, or a single letter like `"X"` to translate them to. Note you could _probably_ get that behaviour right now without changing Biopython by providing a custom table with extra entries like `"AA-" : "X"` (or done at runtime in code via a subclass). But I still don't grasp your intended use for this.

mmokrejs commented:
With biopython-1.85 I get:

```
from Bio import AlignIO
_align = AlignIO.read("tests/inputs/test2.fa", "fasta")
for _seq in _align:
    print(_seq.translate())
...     
Traceback (most recent call last):
  File "<python-input-16>", line 2, in <module>
    print(_seq.translate())
          ~~~~~~~~~~~~~~^^
  File "/usr/lib/python3.13/site-packages/Bio/SeqRecord.py", line 1502, in translate
    self.seq.translate(
    ~~~~~~~~~~~~~~~~~~^
        table=table, stop_symbol=stop_symbol, to_stop=to_stop, cds=cds, gap=gap
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    )
    ^
  File "/usr/lib/python3.13/site-packages/Bio/Seq.py", line 1620, in translate
    _translate_str(str(self), table, stop_symbol, to_stop, cds, gap=gap)
    ~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.13/site-packages/Bio/Seq.py", line 2911, in _translate_str
    raise CodonTable.TranslationError(
        f"Codon '{codon}' is invalid"
    ) from None
Bio.Data.CodonTable.TranslationError: Codon '---' is invalid
>>>
```
Above I used 
[test2.fa.txt](https://github.com/user-attachments/files/20081872/test2.fa.txt)

Editing the sequences and dropping `-` is not a solution although I do it in my code on local variables (not biopython objects) containing codos alone. The point is that biopython is NOT generic enough to `translate()` multiple types of data inputs and provide meaningful output in each scenario. Instead, it chokes. The roadblocks installed were deemed for safety but they just kick-off the user with no way to overcome the "false" errors. That is why there should be multiple optional arguments so that user can in several scenarios overcome the roadblocks. To translate a perfectly pure `[ATGCN]`-containing sequences there are hunderds of tools. But if you want to parse some input files from various sources, you end end up with immutable sequence records. Editing them is a nightmare, merely you end up with a new sequence object. It is largely inefficient because in my case I wanted to parse rather large FASTA files and translate them and do additional processing with both, the nucleic acid sequence and with the protein sequence. I am surprised addition of two optional arguments requires such justification.

@mdehoon I do use `AlignIO.read()` already at https://github.com/host-patho-evo/mutation_scatter_plot/blob/main/calculate_codon_frequencies.py#L148

@peterjc The approach _by providing a custom table with extra entries like "AA-" : "X" (or done at runtime in code via a subclass)_ is crazily complicated. Why should one create a 5^3 table with all codon with one or two dashes in any position? Simply, translate should translate `AA-` as `X` and `A-A` as well. There are a few cases when one could translate `CTN` into `L`, `CGN` into `R`, `TCN` into `S`, `GGN` into `G`, `GCN` into `A`, `GTN` into `V`, `ACN` into `T` and `CCN` into `P`. I do not add that but it would be handy extension allowing me to work around a few sequencing errors in the data.

peterjc commented:
The last point already works as you hoped:

```pycon
>>> from Bio.Seq import Seq
>>> Seq("CTN").translate()
Seq('L')
```

What I think was your main point:

> The point is that biopython is NOT generic enough to translate() multiple types of data inputs and provide meaningful output in each scenario. Instead, it chokes.

Rather I would say when asked to translate without a meaningful interpretation, it gives an error. This seems appropriate to me, *"In the face of ambiguity, refuse the temptation to guess."* came to mind, but other entries in the [Zen of Python](https://peps.python.org/pep-0020/) seem relevant too.

I evidentially still don't understand your use-case here.

peterjc commented:
Another thought, since you want 'AT-' translated to 'X', would `.replace("-", "N").translate()` or maybe `.replace("---", "").replace("-", "N").translate()` suffice?

[We already established `.replace("-", "").translate()` doesn't do what you want]

peterjc commented:
I'd considered `.replace("---", "???").replace("-", "N").replace("???", "---")` too, but the '---' it preserves may not be in phase. For example I _think_ the requested behavior here is to translate 'AA---A' to 'XX'.

[I think that input would be suggestive of a naive alignment which wasn't codon aware, and would want the current error behaviour here.]

mmokrejs commented:
> I'd considered `.replace("---", "???").replace("-", "N").replace("???", "---")` too, but the '---' it preserves may not be in phase. 

@peterjc  Exactly, my example sequence starts in frame +1 but it could be +2 or +3 so it is not very informative to realize the issue. But if you consider `---` somewhere inside the seuqence in frame +3 you would immediately get a wrong result. Indeed, in https://github.com/host-patho-evo/mutation_scatter_plot/blob/main/tests/inputs/test2.fasta#L18 I added a real example of BA2.86 Covid-19 mutation where 3 nucleotides are DELeted but tjose are `pos2, pos3, pos1` so although it results in a 1 amino acid residue being deleted at the DNA level 3nt get DELeted from two underlying codons.

> For example I _think_ the requested behavior here is to translate 'AA---A' to 'XX'.

@peterjc Exactly, but please note intentionally I added the .translate(..., `ignore_gaps=True`) which would internally in biopython perform `.replace('-','')`. Because sometimes it is helpful to try an alternative way (actually third way because the default is current biopython behavior).

Let me say it once again, my patch provides two additional ways how to translate sequences (without need to edit the sequences if tehy just get translated and deleted from memory), while the current approach remains as the default, dying often or raising zillions of warning that codons are not a multiple of three.

> [I think that input would be suggestive of a naive alignment which wasn't codon aware, and would want the current error behaviour here.]

Here I am short with my English. The FASTA files I provide now are manually polished alignments in a way that if possible, nucleotides are in triplets (codons). If that is not possible, an "offending" nucleotide (possible sequencing error INSertion) is in a new codon (e.g. `--A`). By this approach, the sequences can be translated as a whole in frame and eventual INSertion due to sequencing errors breaking the reading frame do not ruin the translate() result with premature `*`. The biological reasoning is the fact that if the sequences are from a protein coding region there were not frame-breaking mutation in the live virus. So any inDel breaking a reading frame in the particular sequence reason is clearly a sequencing error. That is why I wish to get rather na extra `X` but the rest aligned properly. The downside is that I need to manually polish the alignment and widen it with padding symbols.

I admit that initially in my code I walked the "reference" sequence with a window of 3 characters and if any of them was a `-` I fetched one more character. I did not re-align the InDels manually, it did not matter if there is `--A` or `A--` or `-A-`. Once I had a three [ATGCN] chars I translated it. Then the same column-width I translated for each sample sequence (member of the alignment). It worked quite well but after I faced the INSertions in the sample NGS reads (contrary to the reference) and also after I faced the DELetions partially spanning multiple codons (the BA2.86 case above spanning S protein codons 482 and 483) I gave up and changed the code to require alignment where user must re-align the InDels in a way that if possible, all codon consist of three [ATGCN] and gaps consist of three `---`. If that is not possible, partial codons (single- or double-nucleotide INSertions) must be padded with `-` to the length of three. By this approach, I can make "visible"/"actionable" INSertion events in the sample. Until that, only DELetion events were "visible"/"actionable".

I did not check published analyses but it seems to pipelines using gofasta did not realize INSertions (3, 6, 9, 12, ... nt wide) in the sample sequences due to that:
https://github.com/virus-evolution/gofasta/issues/55

For your curiosity, I put together in the past some links to other tools which did not fit my needs:
https://github.com/virus-evolution/gofasta/pull/50

I think I might have inspired somebody providing a pull request partly addressing issues with gofasta but it did not get merged: https://github.com/virus-evolution/gofasta/pull/56

mdehoon commented:
> Note "---" already does translate as "-":
```python
>>> from Bio.Seq import Seq
>>> Seq("---").translate()
Seq('-')
```

It seems that Biopython is not consistent here:
```python
>>> from Bio.Seq import Seq, translate
>>> Seq("---").translate()
Seq('-')
>>> translate(Seq("---"))
Seq('-')
>>> translate("---")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/mdehoon/.venv/lib/python3.12/site-packages/Bio/Seq.py", line 3018, in translate
    return _translate_str(sequence, table, stop_symbol, to_stop, cds, gap=gap)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/mdehoon/.venv/lib/python3.12/site-packages/Bio/Seq.py", line 2911, in _translate_str
    raise CodonTable.TranslationError(
Bio.Data.CodonTable.TranslationError: Codon '---' is invalid
```

peterjc commented:
We should update the translate function to handle '---' too.

peterjc commented:
Is this useful?

```python
from Bio.Data.CodonTable import (
    AmbiguousForwardTable,
    ambiguous_generic_by_id,
    TranslationError,
)
from Bio.Seq import translate


class GappyAmbiguousForwardTable(AmbiguousForwardTable):
    def __getitem__(self, codon):
        try:
            return AmbiguousForwardTable.__getitem__(self, codon)
        except KeyError:
            if "---" == codon:
                self._cache[codon] = "-"
                return "-"
            elif "-" in codon:
                # Might want to check rest of characters in codon
                # are valid nucleotides?
                self._cache[codon] = "!"
                return "!"
            else:
                raise
        except TranslationError:
            self._cache[codon] = "X"
            return "X"


# Monkey-patch NCBI codon table to use gappy translation
# (should perhaps use a copy of the table here?)
gappy_table = ambiguous_generic_by_id[1]
gappy_table.forward_table = GappyAmbiguousForwardTable(
    gappy_table.forward_table.forward_table,
    gappy_table.forward_table.ambiguous_nucleotide,
    gappy_table.forward_table.ambiguous_protein,
)

assert gappy_table.forward_table["NAN"] == "X"
assert gappy_table.forward_table["AA-"] == "!"
assert gappy_table.forward_table["---"] == "-"

assert translate("---", gappy_table) == "-"
assert translate("AA---A", gappy_table) == "!!"
assert translate("TANAA------ANST", gappy_table) == "X!-!X"
```

mmokrejs commented:
> Is this useful?

```
>>> translate("---", gappy_table)
'-'
>>> translate("-A-", gappy_table)
'!'
>>> translate("-W-", gappy_table)
'!'
>>> translate("-O-", gappy_table)
'!'
>>> translate("-.-", gappy_table)
'!'
>>> translate("---ATGatg", gappy_table)
'-MM'
>>> translate("---ATGat-", gappy_table)
'-M!'
>>> translate("AAA---ATGat-", gappy_table)
'K-M!'
>>> translate("AA---ATGat-", gappy_table)
/usr/lib/python3.13/site-packages/Bio/Seq.py:2879: BiopythonWarning: Partial codon, len(sequence) not a multiple of three. Explicitly trim the sequence or add trailing N before translation. This may become an error in future.
  warnings.warn(
'!!*'
>>> translate("AA----ATGat-", gappy_table)
'!-M!'
>>> translate("AA----ATGa4-", gappy_table)
'!-M!'
>>> translate("AA----ATG44-", gappy_table)
'!-M!'
>>>
```

Not bad, although I would prefer `X` instead of `!` for the purpose of debugging it is handy for sure. The main objection is that for the below case it should rather fill up the trailing missing `-` instead of Warning:

```
>>> translate("AA---ATGat-", gappy_table)
/usr/lib/python3.13/site-packages/Bio/Seq.py:2879: BiopythonWarning: Partial codon, len(sequence) not a multiple of three. Explicitly trim the sequence or add trailing N before translation. This may become an error in future.
  warnings.warn(
'!!*'
```
IMO the above should never become an error. Probably `!!*!` would be preferred here.


```
from Bio import AlignIO
_align = AlignIO.read("tests/inputs/test2.fasta", "fasta")
for _seq in _align:
    print(translate(_seq.seq))
...     
M-FVFLVLLPLVSSQC----VNLITRTQ------SYTNSFTRGVYYPDKVFRSSVLHSTQDLFLPFFSNVTWFHAI----SGTNGTKRFDNPVLPFNDGVYFASTEKSNIIRGWIFGTTLDSKTQSLLIVNNATNVVIKVCEFQFCNDPFLDV-YYHKNNKSWMESEFRVYSSANNCTFEYVSQPFLMDLEGKQGNFKNLREFVFKNIDGYFKIYSKHTPINL---GR---DLPQGFSALEPLVDLPIGINITRFQTLLALHRSYLTPGDSSSGWTAGAAAYYVGYLQPRTFLLKYNENGTITDAVDCALDPLSETKCTLKSFTVEKGIYQTSNFRVQPTESIVRFPNITNLCPFDEVFNATRFASVYAWNRKRISNCVADYSVLYNFAPFFAFKCYGVSPTKLNDLCFTNVYADSFVIRGNEVSQIAPGQTGNIADYNYKLPDDFTGCVIAWNSNKLDSKVGGNYNYRYRLFRKSNLKPFERDISTEIYQAGNKPCNGVAGVNCYFPLQSYGFRPTYGVGHQPYRVVVLSFELLHAPATVCGPKKSTNLVKNKCVNFNFNGLTGTGVLTESNKKFLPFQQFGRDIADTTDAVRDPQTLEILDITPCSFGGVSVITPGTNTSNQVAVLYQGVNCTEVPVAIHADQLTPTWRVYSTGSNVFQTRAGCLIGAEYVNNSYECDIPIGAGICASYQTQTKSHRRARSVASQSIIAYTMSLGAENSVAYSNNSIAIPTNFTISVTTEILPVSMTKTSVDCTMYICGDSTECSNLLLQYGSFCTQLKRALTGIAVEQDKNTQEVFAQVKQIYKTPPIKYFGGFNFSQILPDPSKPSKRSFIEDLLFNKVTLADAGFIKQYGDCLGDIAARDLICAQKFNGLTVLPPLLTDEMIAQYTSALLAGTITSGWTFGAGAALQIPFAMQMAYRFNGIGVTQNVLYENQKLIANQFNSAIGKIQDSLSSTASALGKLQDVVNHNAQALNTLVKQLSSKFGAISSVLNDILSRLDKVEAEVQIDRLITGRLQSLQTYVTQQLIRAAEIRASANLAATKMSECVLGQSKRVDFCGKGYHLMSFPQSAPHGVVFLHVTYVPAQEKNFTTAPAICHDGKAHFPREGVFVSNGTHWFVTQRNFYEPQIITTDNTFVSGNCDVVIGIVNNTVYDPLQPELDSFKEELDKYFKNHTSPDVDLGDISGINASVVNIQKEIDRLNEVAKNLNESLIDLQELGKYEQYIKWPWYIWLGFIAGLIAIVMVTIMLCCMTSCCSCLKGCCSCGSCCKFDEDDSEPVLKGVKLHYT*
M-FVFLVLLPLVSSQC----VNLITRTQ------SYTNSFTRGVYYPDKVFRSSVLHSTQDLFLPFFSNVTWFHAIH--VSGTNGTKRFDNPVLPFNDGVYFASTEKSNIIRGWIFGTTLDSKTQSLLIVNNATNVVIKVCEFQFCNDPFLDV-YYHKNNKSWMESEFRVYSSANNCTFEYVSQPFLMDLEGKQGNFKNLREFVFKNIDGYFKIYSKHTPINL---GR---DLPQGFSALEPLVDLPIGINITRFQTLLALHRSYLTPGDSSSGWTAGAAAYYVGYLQPRTFLLKYNENGTITDAVDCALDPLSETKCTLKSFTVEKGIYQTSNFRVQPTESIVRFPNITNLCPFDEVFNATRFASVYAWNRKRISNCVADYSVLYNFAPFFAFKCYGVSPTKLNDLCFTNVYADSFVIRGNEVSQIAPGQTGNIADYNYKLPDDFTGCVIAWNSNKLDSKVGGNYNYLYRLFRKSNLKPFERDISTEIYQAGNKPCNGVAGFNCYFPLRSYGFRPTYGVGHQPYRVVVLSFELLHAPATVCGPKKSTNLVKNKCVNFNFNGLTGTGVLTESNKKFLPFQQFGRDIADTTDAVRDPQTLEILDITPCSFGGVSVITPGTNTSNQVAVLYQGVNCTEVPVAIHADQLTPTWRVYSTGSNVFQTRAGCLIGAEYVNNSYECDIPIGAGICASYQTQTKSHRRARSVASQSIIAYTMSLGAENSVAYSNNSIAIPTNFTISVTTEILPVSMTKTSVDCTMYICGDSTECSNLLLQYGSFCTQLKRALTGIAVEQDKNTQEVFAQVKQIYKTPPIKYFGGFNFSQILPDPSKPSKRSFIEDLLFNKVTLADAGFIKQYGDCLGDIAARDLICAQKFNGLTVLPPLLTDEMIAQYTSALLAGTITSGWTFGAGAALQIPFAMQMAYRFNGIGVTQNVLYENQKLIANQFNSAIGKIQDSLSSTASALGKLQDVVNHNAQALNTLVKQLSSKFGAISSVLNDILSRLDKVEAEVQIDRLITGRLQSLQTYVTQQLIRAAEIRASANLAATKMSECVLGQSKRVDFCGKGYHLMSFPQSAPHGVVFLHVTYVPAQEKNFTTAPAICHDGKAHFPREGVFVSNGTHWFVTQRNFYEPQIITTDNTFVSGNCDVVIGIVNNTVYDPLQPELDSFKEELDKYFKNHTSPDVDLGDISGINASVVNIQKEIDRLNEVAKNLNESLIDLQELGKYEQYIKWPWYIWLGFIAGLIAIVMVTIMLCCMTSCCSCLKGCCSCGSCCKFDEDDSEPVLKGVKLHYT*
M-FVFLVLLPLVSSQC----VNLTTRTQ---LPPAYTNSFTRGVYYPDKVFRSSVLHSTQDLFLPFFSNVTWFHAI----SGTNGTKRFDNPVLPFNDGVYFASTEKSNIIRGWIFGTTLDSKTQSLLIVNNATNVVIKVCEFQFCNDPFLGV--YHKNNKSWMESEFRVYSSANNCTFEYVSQPFLMDLEGKQGNFKNLREFVFKNIDGYFKIYSKHTPINL---VR---DLPQGFSALEPLVDLPIGINITRFQTLLALHRSYLTPGDSSSGWTAGAAAYYVGYLQPRTFLLKYNENGTITDAVDCALDPLSETKCTLKSFTVEKGIYQTSNFRVQPTESIVRFPNITNLCPFGEVFNATRFASVYAWNRKRISNCVADYSVLYNSASFSTFKCYGVSPTKLNDLCFTNVYADSFVIRGDEVRQIAPGQTGKIADYNYKLPDDFTGCVIAWNSNNLDSKVGGNYNYLYRLFRKSNLKPFERDISTEIYQAGSTPCNGVEGFNCYFPLQSYGFQPTYGVGYQPYRVVVLSFELLHAPATVCGPKKSTNLVKNKCVNFNFNGLTGTGVLTESNKKFLPFQQFGRDIDDTTDAVRDPQTLEILDITPCSFGGVSVITPGTNTSNQVAVLYQGVNCTEVPVAIHADQLTPTWRVYSTGSNVFQTRAGCLIGAEHVNNSYECDIPIGAGICASYQTQTNSHRRARSVASQSIIAYTMSLGAENSVAYSNNSIAIPINFTISVTTEILPVSMTKTSVDCTMYICGDSTECSNLLLQYGSFCTQLNRALTGIAVEQDKNTQEVFAQVKQIYKTPPIKDFGGFNFSQILPDPSKPSKRSFIEDLLFNKVTLADAGFIKQYGDCLGDIAARDLICAQKFNGLTVLPPLLTDEMIAQYTSALLAGTITSGWTFGAGAALQIPFAMQMAYRFNGIGVTQNVLYENQKLIANQFNSAIGKIQDSLSSTASALGKLQDVVNQNAQALNTLVKQLSSNFGAISSVLNDILARLDKVEAEVQIDRLITGRLQSLQTYVTQQLIRAAEIRASANLAATKMSECVLGQSKRVDFCGKGYHLMSFPQSAPHGVVFLHVTYVPAQEKNFTTAPAICHDGKAHFPREGVFVSNGTHWFVTQRNFYEPQIITTHNTFVSGNCDVVIGIVNNTVYDPLQPELDSFKEELDKYFKNHTSPDVDLGDISGINASVVNIQKEIDRLNEVAKNLNESLIDLQELGKYEQYIKWPWYIWLGFIAGLIAIVMVTIMLCCMTSCCSCLKGCCSCGSCCKFDEDDSEPVLKGVKLHYT*
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------LHRSYLTPGDSSSGWTAGAAAYYVGYLQPRTFLLKYNENGTITDAVDCALDPLSETKCTLKSFTVEKGIYQTSNFRVQPTESIVRFPNITNLCPFGEVFNATRFASVYAWNRKRISNCVADYSVLYNSASFSTFKCYGVSPTKLNDLCFTNVYADSFVIRGDEVRQIAPGQTGKIADYNYKLPDDFTGCVIAWNSNNLDSKVGGNYNYRYRLFRKSNLKPFERDISTEIYQAGSKPCNGVEGFNCYFPLQSYGFQPTNGVGYQPYRVVVLSFELLHAPATVCGPKKSTNLVKNKCVNFNFNGLTGTGVLTESNKKFLPFQQFGRDIADTTDAVRDPQTLEILDITPCSFGGVSVITPGTNTSNQVAVLYQGVNCTEVPVAIHADQLTPTWRVYSTGSNVFQTRAGCLIGAEHVNNSYECDIPIGAGICASYQTQTNSRRRARSVASQSIIAYTMSLGAENSVAYSNNSIAIPTNFTISVTTEILPVSMTKTSVDCTMYICGDSTECSNLLLQYGSFCTQLNRALTGIAVEQDKNTQEVFAQVKQIYKTPPIKDFGGFNFSQILPDPSKPSKRSFIEDLLFNKVTLADAGFIKQYGDCLGDIAARDLICAQKFNGLTVLPPLLTDEMIAQYTSALLAGTITSGWTFGAGAALQIPFAMQMAYRFNGIGVTQNVLYENQKLIANQFNSAIGKIQDSLSSTASALGKLQNVVNQNAQALNTLVKQLSSNFGAISSVLNDILSRLDKVEAEVQIDRLITGRLQSLQTYVTQQLIRAAEIRASANLAATKMSECVLGQSKRVDFCGKGYHLMSFPQSAPHGVVFLHVTYVPAQEKNFTTAPAICHDGKAHFPREGVFVSNGTHWFVTQRNFYEPQIITTDNTFVSGNCDVVIGIVNNTVYDPLQPELDSFKEELDKYFKNHTSPDVDLGDISGINASVVNIQKEIDRLNEVAKNLNESLIDLQELGKYEQYIKWPWYIWLGFIAGLIAIVMVTIMLCCMTSCCSCLKGCCSCGSCCKFDEDDSEPVLKGVKLHYT*
M-FVFLVLLPLVSSQC----VNLITTTQ------SYTNSFTRGVYYPDKVFRSSVLHLTQDLFLPFFSNVTWFHAI----SGTNGTKRFDNPVLPFNDGVYFASTEKSNIIRGWIFGTTLDSKTQSLLIVNNATNVFIKVCEFQFCNDPFLDV--YHKNNKSWMESESGVYSSANNCTFEYVSQPFLMDLEGKQGNFKNLREFVFKNIDGYFKIYSKHTPI-I---GR---DFPQGFSALEPLVDLPIGINITRFQTLLALNRSYLTPGDSSSGWTAGAADYYVGYLQPRTFLLKYNENGTITDAVDCALDPLSETKCTLKSFTVEKGIYQTSNFRVQPTESIVRFPNVTNLCPFHEVFNATRFASVYAWNRTRISNCVADYSVLYNFAPFFAFKCYGVSPTKLNDLCFTNVYADSFVIKGNEVSQIAPGQTGNIADYNYKLPDDFTGCVIAWNSNKLDSKHSGNYDYWYRSFRKSKLKPFERDISTEIYQAGNKPCKG-KGPNCYFPLQSYGFRPTYGVGHQPYRVVVLSFELLHAPATVCGPKKSTNLVKNKCVNFNFNGLTGTGVLTKSNKKFLPFQQFGRDIVDTTDAVRDPQTLEILDITPCSFGGVSVITPGTNTSNQVAVLYQGVNCTEVSVAIHADQLTPTWRVYSTGSNVFQTRAGCLIGAEYVNNSYECDIPIGAGICASYQTQTKSRRRARSVASQSIIAYTMSLGAENSVAYSNNSIAIPTNFTISVTTEILPVSMTKTSVDCTMYICGDSTECSNLLLQYGSFCTQLKRALTGIAVEQDKNTQEVFAQVKQIYKTPPIKYFGGFNFSQILPDPSKPSKRSFIEDLLFNKVTLADAGFIKQYGDCLGDIAARDLICAQKFNGLTVLPPLLTDEMIAQYTSALLAGTITSGWTFGAGAALQIPFAMQMAYRFNGIGVTQNVLYENQKLIANQFNSAIGKIQDSLFSTASALGKLQDVVNHNAQALNTLVKQLSSKFGAISSVLNDILSRLDKVEAEVQIDRLITGRLQSLQTYVTQQLIRAAEIRASANLAATKMSECVLGQSKRVDFCGKGYHLMSFPQSAPHGVVFLHVTYVPAQEKNFTTAPAICHDGKAHFPREGVFVSNGTHWFVTQRNFYEPQIITTDNTFVSGNCDVVIGIVNNTVYDPLQLELDSFKEELDKYFKNHTSPDVDLGDISGINASVVNIQKEIDRLNEVAKNLNESLIDLQELGKYEQYIKWPWYIWLGFIAGLIAIVMVTIMLCCMTSCCSCLKGCCSCGSCCKFDEDDSEPVLKGVKLHYT*
M-FVFLVLLPLVSSQCVMPLFNLITTTQ------SYTNSFTRGVYYPDKVFRSSVLHLTQDLFLPFFSNVTWFHAI----SGTNGTKRFDNPVLPFNDGVYFASTEKSNIIRGWIFGTTLDSKTQSLLIVNNATNVFIKVCEFQFCNDPFLDV--YHKNNKSWMESESGVYSSANNCTFEYVSQPFLMDLEGKQGNFKNLREFVFKNIDGYFKIYSKHTPI-I---GR---DFPQGFSALEPLVDLPIGINITRFQTLLALNRSYLTPGDSSSGWTAGAADYYVGYLQPRTFLLKYNENGTITDAVDCALDPLSETKCTLKSFTVEKGIYQTSNFRVQPTESIVRFPNVTNLCPFHEVFNATRFASVYAWNRTRISNCVADYSVLYNFAPFFAFKCYGVSPTKLNDLCFTNVYADSFVIKGNEVSQIAPGQTGNIADYNYKLPDDFTGCVIAWNSNKLDSKHSGNYDYWYRSFRKSKLKPFERDISTEIYQAGNKPCKG-KGPNCYFPLQSYGFRPTYGVGHQPYRVVVLSFELLHAPATVCGPKKSTNLVKNKCVNFNFNGLTGTGVLTKSNKKFLPFQQFGRDIVDITDAVRDPQTLEILDITPCSFGGVSVITPGTNTSNQVAVLYQGVNCTEVSVAIHADQLTPTWRVYSTGSNVFQTRAGCLIGAEYVNNSYECDIPIGAGICASYQTQTKSRRRARSVASQSIIAYTMSLGAENSVAYSNNSIAIPTNFTISVTTEILPVSMTKTSVDCTMYICGDSTECSNLLLQYGSFCTQLKRALTGIAVEQDKNTQEVFAQVKQIYKTPPIKYFGGFNFSQILPDPSKPSKRSFIEDLLFNKVTLADAGFIKQYGDCLGDIAARDLICAQKFNGLTVLPPLLTDEMIAQYTSALLAGTITSGWTFGAGAALQIPFAMQMAYRFNGIGVTQNVLYENQKLIANQFNSAIGKIQDSLFSTASALGKLQDVVNHNAQALNTLVKQLSSKFGAISSVLNDILSRLDKVEAEVQIDRLITGRLQSLQTYVTQQLIRAAEIRASANLAATKMSECVLGQSKRVDFCGKGYHLMSFPQSAPHGVVFLHVTYVPAQEKNFTTAPAICHDGKAHFPREGVFVSNGTHWFVTQRNFYEPQIITTDNTFVSGNCDVVIGIVNNTVYDPLQLELDSFKEELDKYFKNHTSPDVDLGDISGINASVVNIQKEIDRLNEVAKNLNESLIDLQELGKYEQYIKWPWYIWLGFIAGLIAIVMVTIMLCCMTSCCSCLKGCCSCGSCCKFDEDDSEPVLKGVKLHYT*
M-FVFLVLLPLVSSQCVMPLFNLITTTQ------SYTN-FTRGVYYPDKVFRSSVLHLTQDLFLPFFSNVTWFHAI----SGTNGTKRFDNPVLPFNDGVYFASTEKSNIIRGWIFGTTLDSKTQSLLIVNNATNVFIKVCEFQFCNDPFLDV--YQKNNKSWMESESGVYSSANNCTFEYVSQPFLMDLEGKQGNFKNLREFVFKNIDGYFKIYSKHTPI-I---GR---DFPQGFSALEPLVDLPIGINITRFQTLLALNRSYLTPGDSSSGWTAGAADYYVGYLQPRTFLLKYNENGTITDAVDCALDPLSETKCTLKSFTVEKGIYQTSNFRVQPTESIVRFPNVTNLCPFHEVFNATTFASVYAWNRTRISNCVADYSVLYNFAPFFAFKCYGVSPTKLNDLCFTNVYADSFVIKGNEVSQIAPGQTGNIADYNYKLPDDFTGCVIAWNSNKLDSKHSGNYDYWYRSLRKSKLKPFERDISTEIYQAGNKPCKG-KGPNCYFPLQSYGFRPTYGVGHQPYRVVVLSFELLHAPATVCGPKKSTNLVKNKCVNFNFNGLTGTGVLTKSNKKFLPFQQFGRDIVDTTDAVRDPQTLEILDITPCSFGGVSVITPGTNTSNQVAVLYQGVNCTEVSVAIHADQLTPTWRVYSTGSNVFQTRAGCLIGAEYVNNSYECDIPIGAGICASYQTQTKSRRRARSVASQSIIAYTMSLGAENSVAYSNNSIAIPTNFTISVTTEILPVSMTKTSVDCTMYICGDSTECSNLLLQYGSFCTQLKRALTGIAVEQDKNTQEVFAQVKQIYKTPPIKYFGGFNFSQILPDPSKPSKRSFIEDLLFNKVTLADAGFIKQYGDCLGDIAARDLICAQKFNGLTVLPPLLTDEMIAQYTSALLAGTITSGWTFGAGAALQIPFAMQMAYRFNGIGVTQNVLYENQKLIANQFNSAIGKIQDSLFSTASALGKLQDVVNHNAQALNTLVKQLSSKFGAISSVLNDILSRLDKVEAEVQIDRLITGRLQSLQTYVTQQLIRAAEIRASANLAATKMSECVLGQSKRVDFCGKGYHLMSFPQSAPHGVVFLHVTYVPAQEKNFTTAPAICHDGKAHFPREGVFVSNGTHWFLTQRNFYEPQIITTDNTFVSGNCDVVIGIVNNTVYDPLQLELDSFKEELDKYFKNHTSPDVDLGDISGINASVVNIQKEIDRLNEVAKNLNESLIDLQELGKYEQYIKWPWYIWLGFIAGLIAIVMVTIMLCCMTSCCSCLKGCCSCGSCCKFDEDDSEPVLKGVKLHYT*
M-FVFLVLLPLVSSQC----VNLTTRTQ---LPPAYTNSFTRGVYYPDKVFRSSVLHSTQDLFLPFFSNVTWFHAI----SGTNGTKRFDNPVLPFNDGVYFASTEKSNIIRGWIFGTTLDSKTQSLLIVNNATNVVIKVCEFQFCNDPFLGV--YHKNNKSWMESEFRVYSSANNCTFEYVSQPFLMDLEGKQGNFKNLREFVFKNIDGYFKIYSKHTPINL---VR---DLPQGFSALEPLVDLPIGINITRFQTLLALHRSYLTPGDSSSGWTAGAAAYYVGYLQPRTFLLKYNENGTITDAVDCALDPLSETKCTLKSFTVEKGIYQTSNFRVQPTESIVRFPNITNLCPFGEVFNATRFASVYAWNRKRISNCVADYSVLYNSASFSTFKCYGVSPTKLNDLCFTNVYADSFVIRGDEVRQIAPGQTGKIADYNYKLPDDFTGCVIAWNSNNLDSKVGGNYNYLYRLFRKSNLKPFERDISTEIYQAGSTPCNGVEGFNCYFPLQSYGFQPTYGVGYQPYRVVVLSFELLHAPATVCGPKKSTNLVKNKCVNFNFNGLTGTGVLTESNKKFLPFQQFGRDIDDTTDAVRDPQTLEILDITPCSFGGVSVITPGTNTSNQVAVLYQGVNCTEVPFAIHADQLTPTWRVYSTGSNVFQTRAGCLIGAEHVNNSYECDIPIGAGICASYQTQTNSHRRARSVASQSIIAYTMSLGAENSVAYSNNSIAIPINFTISVTTEILPVSMTKTSVDCTMYICGDSTECSNLLLQYGSFCTQLNRALTGIAVEQDKNTQEVFAQVKQIYKTPPIKDFGGFNFSQILPDPSKPSKRSFIEDLLFNKVTLADAGFIKQYGDCLGDIAARDLICAQKFNGLTVLPPLLTDEMIAQYTSALLAGTITSGWTFGAGAALQIPFAMQMAYRFNGIGVTQNVLYENQKLIANQFNSAIGKIQDSLSSTASALGKLQDVVNQNAQALNTLVKQLSSNFGAISSVLNDILARLDKVEAEVQIDRLITGRLQSLQTYVTQQLIRAAEIRASANLAATKMSECVLGQSKRVDFCGKGYHLMSFPQSAPHGVVFLHVTYVPAQEKNFTTAPAICHDGKAHFPREGVFVSNGTHWFVTQRNFYEPQIITTHNTFVSGNCDVVIGIVNNTVYDPLQPELDSFKEELDKYFKNHTSPDVDLGDISGINASVVNIQKEIDRLNEVAKNLNESLIDLQELGKYEQYIKWPWYIWLGFIAGLIAIVMVTIMLCCMTSCCSCLKGCCSCGSCCKFDEDDSEPVLKGVKLHYT*
M-FVFLVLLPLVSSQC----VNLTTRTQ---LPPAYTNSFTRGVYYPDKVFRSSVLHSTQDLFLPFFSNVTWFHAI----SGTNGTKRFDNPVLPFNDGVYFASTEKSNIIRGWIFGTTLDSKTQSLLIVNNATNVVIKVCEFQFCNDPFLGV--YHKNNKSWMESEFRVYSSANNCTFEYVSQPFLMDLEGKQGNFKNLREFVFKNIDGYFKIYSKHTPINL---VR---DLPQGFSALEPLVDLPIGINITRFQTLLALHRSYLTPGDSSSGWTAGAAAYYVGYLQPRTFLLKYNENGTITDAVDCALDPLSETKCTLKSFTVEKGIYQTSNFRVQPTESIVRFPNITNLCPFGEVFNATRFASVYAWNRKRISNCVADYSVLYNSASFSTFKCYGVSPTKLNDLCFTNVYADSFVIRGDEVRQIAPGQTGKIADYNYKLPDDFTGCVIAWNSNNLDSKVGGNYNYLYRLFRKSNLKPFERDISTEIYQAGSTPCN!!EGFNCYFPLQSYGFQPTYGVGYQPYRVVVLSFELLHAPATVCGPKKSTNLVKNKCVNFNFNGLTGTGVLTESNKKFLPFQQFGRDIDDTTDAVRDPQTLEILDITPCSFGGVSVITPGTNTSNQVAVLYQGVNCTEVPFAIHADQLTPTWRVYSTGSNVFQTRAGCLIGAEHVNNSYECDIPIGAGICASYQTQTNSHRRARSVASQSIIAYTMSLGAENSVAYSNNSIAIPINFTISVTTEILPVSMTKTSVDCTMYICGDSTECSNLLLQYGSFCTQLNRALTGIAVEQDKNTQEVFAQVKQIYKTPPIKDFGGFNFSQILPDPSKPSKRSFIEDLLFNKVTLADAGFIKQYGDCLGDIAARDLICAQKFNGLTVLPPLLTDEMIAQYTSALLAGTITSGWTFGAGAALQIPFAMQMAYRFNGIGVTQNVLYENQKLIANQFNSAIGKIQDSLSSTASALGKLQDVVNQNAQALNTLVKQLSSNFGAISSVLNDILARLDKVEAEVQIDRLITGRLQSLQTYVTQQLIRAAEIRASANLAATKMSECVLGQSKRVDFCGKGYHLMSFPQSAPHGVVFLHVTYVPAQEKNFTTAPAICHDGKAHFPREGVFVSNGTHWFVTQRNFYEPQIITTHNTFVSGNCDVVIGIVNNTVYDPLQPELDSFKEELDKYFKNHTSPDVDLGDISGINASVVNIQKEIDRLNEVAKNLNESLIDLQELGKYEQYIKWPWYIWLGFIAGLIAIVMVTIMLCCMTSCCSCLKGCCSCGSCCKFDEDDSEPVLKGVKLHYT*
>>>
```

In the above testcases note several variants of the Covid-19 S protein 480-487 aa region:

1. `NGVEGFN` # covid-19 Wuhan reference region
2. `N!!EGFN` # technically correct alignment but leaving --- spanning two codons which does not match reality because ribosome simply "fetched" next nucleotide and the resulting protein is simply lacking one amino acid
3. `KG-KGPN` # the only biologically meaninful translation but manual InDel realignment is needed to yield codon-aligned so that the output matches what a ribosome producted in reality

I would certainly be fine with this if this happens inside biopython internally. I still think optional arguments for the translate() fucntion are way to go, as they are three situations (default, remove paddings on the fly, translated padded sequence). Now maybe one more, whether user wants `X` or `!` for partial codons).

mdehoon commented:
What if you split up the sequence into codons?
```python
>>> from Bio.Seq import translate
>>> sequence = "AA---ATGat-"
>>> codons = (sequence[i:i+3] for i in range(0, len(sequence), 3))
>>> codons = ("NNN" if "-" in codon and codon != "---" else codon for codon in codons)
>>> aa = "".join(translate(codon) for codon in codons)
>>> aa
'XX*X'
```

mmokrejs commented:
> What if you split up the sequence into codons?
> 
> ```python
> >>> from Bio.Seq import translate
> >>> sequence = "AA---ATGat-"
> >>> codons = (sequence[i:i+3] for i in range(0, len(sequence), 3))
> >>> codons = ("NNN" if "-" in codon and codon != "---" else codon for codon in codons)
> >>> aa = "".join(translate(codon) for codon in codons)
> >>> aa
> 'XX*X'
> ```

The above works around, almost. But for `sequence = "---AA---ATGat-"` it hits a road-block as usually (in installed fresh biopython-1.85 via pip to be sure I have vanilla setup):

```
>>> sequence = "---AA---ATGat-"
>>> codons = (sequence[i:i+3] for i in range(0, len(sequence), 3))
>>> codons = ("NNN" if "-" in codon and codon != "---" else codon for codon in codons)
>>> aa = "".join(translate(codon) for codon in codons)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<stdin>", line 1, in <genexpr>
  File "/home/mmokrejs/virtualenv/test/lib/python3.11/site-packages/Bio/Seq.py", line 3018, in translate
    return _translate_str(sequence, table, stop_symbol, to_stop, cds, gap=gap)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/mmokrejs/virtualenv/test/lib/python3.11/site-packages/Bio/Seq.py", line 2911, in _translate_str
    raise CodonTable.TranslationError(
Bio.Data.CodonTable.TranslationError: Codon '---' is invalid
>>> 
```

mdehoon commented:
@mmokrejs 

> But for sequence = "---AA---ATGat-" it hits a road-block as usually (I installed fresh biopython-1.85 via pip to be sure I have vanilla setup):

I believe @peterjc  is already working to resolve this; see #4994

mmokrejs commented:
> @mmokrejs
> 
> > But for sequence = "---AA---ATGat-" it hits a road-block as usually (I installed fresh biopython-1.85 via pip to be sure I have vanilla setup):
> 
> I believe @peterjc is already working to resolve this; see #4994

@mdehoon Thank you. I had a look into those commits before but I haven't seen the code around line 2923 touched.

I tried now also:

```
from Bio.Seq import translate
sequence = "---AA---ATGat-"
codons = (sequence[i:i+3] for i in range(0, len(sequence), 3))
codons = ("NNN" if "-" in codon and codon != "---" else codon for codon in codons)
"".join(translate(codon, gap='-') for codon in codons)
'XX*X'
```

but that actually deleted the leading '---' on-the-fly, which I did not want to happen. So providing `gap='-'` implies biopython replaces the gap char with an empty string somewhere.

peterjc commented:
Your example is working for me on 1.85 and the master branch:

```pycon
>>> from Bio.Seq import translate
>>> sequence = "---AA---ATGat-"
>>> codons = (sequence[i:i+3] for i in range(0, len(sequence), 3))
>>> codons = ("NNN" if "-" in codon and codon != "---" else codon for codon in codons)
>>> "".join(translate(codon, gap='-') for codon in codons)
'-XX*X'
```

mmokrejs commented:
> Your example is working for me on 1.85 and the master branch:

@peterjc Ah, my fault. I actually performed the  `"".join(translate(codon) for codon in codons)` which fetched from the iterator the fist codon, then retired with `"".join(translate(codon, gap='-') for codon in codons)` which acted on the remainder. Sorry for the noise. I copied the lines manually from my around screen and haven't realized this.
